File: 0136_Single_Number/sol.py
class Solution:
    """
    @param arg1: List
    @return: int
    Time:    O(N)
    Space:   O(N)
    Strategy: XOR
    """
    def solve(self, nums: list[int]) -> int:
        result = 0
        for x in nums:
            result ^= x
        return result

# ================ 🧪 Local Test Suite ======================================
def run_tests():
    sol = Solution()
    test_cases = [
        (( [2,2,1], ), 1 ),
        (( [4,1,2,1,2], ) , 4 ),
        (( [1], ) , 1 ),
        
    ]
    for i, (args, expected) in enumerate(test_cases, 1):
        result = sol.solve(*args)
        status = "✅ PASS" if result == expected else "❌ FAIL"
        print(f"Test {i}: | {status} | Input={args} | Expected={expected} | Got={result}")

if __name__ == "__main__":
    run_tests()

File: 0136_Single_Number/sol.js
/**
 * @param {number[]} nums
 * @return {number}
 * Time:    O(N)
 * Space:   O(N)
 * Strategy: XOR
 */
var func = function(nums) {
    let result = 0;
    for (const x of nums) {
        result ^= x;
    }
    return result;
};
/**
 * Local testing
 */
function test() {
    const testCases = [
        { args: [2,2,1], expected: 1 },
        { args: [4,1,2,1,2], expected: 4 },
        { args: [1], expected: 1 }
    ];
    testCases.forEach(({ args, expected }, index) => {
        // Use spread syntax (...) to pass array elements as separate arguments
        const result = func(args);
        const status = (result === expected) ? "✅ PASS" : "❌ FAIL";
        console.log(
            `${status} | Test #${index + 1}: ` +
            `Input: ${JSON.stringify(args)} | ` +
            `Expected: ${expected} | ` +
            `Result: ${result}`
        );
    });
}
test();
